File: F1Manager2027-Unity/Assets/Scripts/Core/DriverMarket.cs
// DriverMarket.cs — Transfer window: contracts, competitor interest, salary budget.
// Pure C#, no Unity dependencies.
using System;
using System.Collections.Generic;
using System.Linq;

namespace F1Manager
{
    // ─────────────────────────────────────────────────────────────────────────
    // Contract
    // ─────────────────────────────────────────────────────────────────────────

    public class DriverContract
    {
        public int    DriverID;
        public int    TeamID;
        public int    DurationYears;    // 1–3 years
        public int    SalaryMillions;   // Annual salary (M€)
        public int    SeasonStart;      // Season when contract began
        public int    SeasonEnd;        // Season when contract expires (exclusive)
        public bool   HasVetoClause;    // Driver can block a transfer

        public bool IsExpired(int currentSeason) => currentSeason >= SeasonEnd;
        public int  YearsRemaining(int currentSeason) => Math.Max(0, SeasonEnd - currentSeason);

        public string Describe(int currentSeason)
        {
            string veto = HasVetoClause ? " [VETO]" : "";
            int remaining = YearsRemaining(currentSeason);
            string exp = remaining == 0 ? "Expiring" : $"{remaining}yr remaining";
            return $"{exp}, €{SalaryMillions}M/yr{veto}";
        }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Transfer offer
    // ─────────────────────────────────────────────────────────────────────────

    public class TransferOffer
    {
        public int    OfferingTeamID;
        public int    DriverID;
        public int    ProposedSalaryMillions;
        public int    ProposedDurationYears;
        public bool   IsPlayerOffer;   // True = human player sent this

        // Outcome
        public bool   IsAccepted;
        public bool   IsDeclined;
        public string DeclineReason;
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Driver market value
    // ─────────────────────────────────────────────────────────────────────────

    /// <summary>Computed market value for one driver at a given point in time.</summary>
    public class DriverMarketValue
    {
        public int   DriverID;
        public int   BaseValueMillions;     // Annual market rate
        public int   MinAcceptableMillions; // Will not sign below this
        public bool  IsMarketAvailable;     // Not under contract or contract expiring
        public int   InterestingTeamCount;  // How many teams are actively pursuing
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Transfer window state
    // ─────────────────────────────────────────────────────────────────────────

    public class TransferWindowState
    {
        public int  Season;
        public bool IsOpen;  // Transfer window opens at season end / mid-season

        public List<DriverContract>  Contracts = new List<DriverContract>();
        public List<TransferOffer>   PendingOffers = new List<TransferOffer>();
        public List<TransferOffer>   ClosedOffers  = new List<TransferOffer>();

        // Team salary budgets (M€ per year)
        public Dictionary<int, int>  TeamBudgets = new Dictionary<int, int>();
        public Dictionary<int, int>  TeamSpent   = new Dictionary<int, int>(); // Committed salary

        /// <summary>Returns contracts for a specific team.</summary>
        public List<DriverContract> GetTeamContracts(int teamID)
            => Contracts.Where(c => c.TeamID == teamID).ToList();

        /// <summary>Returns the contract for a driver, or null if free agent.</summary>
        public DriverContract GetDriverContract(int driverID)
            => Contracts.FirstOrDefault(c => c.DriverID == driverID);

        /// <summary>Budget remaining for a team after committed salary.</summary>
        public int RemainingBudget(int teamID)
        {
            int budget  = TeamBudgets.TryGetValue(teamID, out int b) ? b : 20;
            int spent   = TeamSpent.TryGetValue(teamID, out int s)   ? s : 0;
            return budget - spent;
        }
    }

    // ─────────────────────────────────────────────────────────────────────────
    // DriverMarket
    // ─────────────────────────────────────────────────────────────────────────

    /// <summary>
    /// Manages driver contracts and the transfer window.
    /// At season end, expired contracts free drivers for transfer.
    /// AI teams bid on available drivers based on their ratings and budget.
    /// Players can make offers within their team's salary budget.
    /// </summary>
    public class DriverMarket
    {
        public TransferWindowState State { get; private set; }
        private readonly System.Random _rng;

        // Salary table (millions per year) indexed by driver pace/racecraft average
        // High-rated drivers demand more; top teams pay premium
        private static readonly (int minRating, int baseSalary)[] _salaryBands =
        {
            (95, 45), // Verstappen-tier: €45M+
            (90, 30), // Hamilton/Leclerc-tier: €30M
            (85, 20), // Solid race winner: €20M
            (80, 14), // Mid-field regular: €14M
            (75, 10), // Lower mid-field: €10M
            (0,   6), // Rookie/lower grid: €6M
        };

        // Team salary budgets (M€ per season — combined for two drivers)
        private static readonly Dictionary<string, int> _teamBudgetBands = new Dictionary<string, int>
        {
            { "top",   90 },   // RedBull, Mercedes, Ferrari
            { "mid",   60 },   // McLaren, Aston
            { "lower", 35 },   // Williams, Haas, Sauber, Alpine
        };

        public DriverMarket(int season = 2027, System.Random rng = null)
        {
            _rng  = rng ?? new System.Random();
            State = new TransferWindowState { Season = season };
            GameDataFactory.Initialize();
            InitialiseContracts();
            InitialiseBudgets();
        }

        // ── Initialisation ────────────────────────────────────────────────────

        private void InitialiseContracts()
        {
            // Give all active drivers a 1–3 year contract at season start
            foreach (var driver in GameDataFactory.Drivers)
            {
                int salary   = CalculateBaseSalary(driver);
                int duration = _rng.Next(1, 4);  // 1–3 years
                bool veto    = salary >= 25;       // Only top drivers have veto clauses

                var contract = new DriverContract
                {
                    DriverID        = driver.id,
                    TeamID          = driver.teamID,
                    DurationYears   = duration,
                    SalaryMillions  = salary,
                    SeasonStart     = State.Season,
                    SeasonEnd       = State.Season + duration,
                    HasVetoClause   = veto,
                };
                State.Contracts.Add(contract);
            }
        }

        private void InitialiseBudgets()
        {
            foreach (var team in GameDataFactory.Teams)
            {
                string tier = team.carPerformance >= 90 ? "top"
                            : team.carPerformance >= 83 ? "mid"
                            : "lower";
                State.TeamBudgets[team.id] = _teamBudgetBands[tier];

                // Committed salary = sum of all driver salaries on this team
                int spent = State.Contracts
                    .Where(c => c.TeamID == team.id)
                    .Sum(c => c.SalaryMillions);
                State.TeamSpent[team.id] = spent;
            }
        }

        // ── Market value helpers ──────────────────────────────────────────────

        /// <summary>Computes market value for a driver at current time.</summary>
        public DriverMarketValue GetMarketValue(int driverID)
        {
            var driver   = GameDataFactory.GetDriver(driverID);
            if (driver == null) return null;

            int   base_  = CalculateBaseSalary(driver);
            var   contract = State.GetDriverContract(driverID);
            bool  free   = contract == null || contract.IsExpired(State.Season);

            // AI interest: teams with budget and an expiring seat
            int interest = State.Contracts.Count(c =>
                c.DriverID != driverID
                && c.IsExpired(State.Season)
                && State.RemainingBudget(c.TeamID) >= base_ * 8 / 10);

            return new DriverMarketValue
            {
                DriverID               = driverID,
                BaseValueMillions      = base_,
                MinAcceptableMillions  = (int)(base_ * 0.75f),
                IsMarketAvailable      = free,
                InterestingTeamCount   = Math.Min(interest, 3),
            };
        }

        private int CalculateBaseSalary(DriverInfo driver)
        {
            int avg = (driver.pace + driver.racecraft) / 2;
            foreach (var (minRating, baseSalary) in _salaryBands)
                if (avg >= minRating) return baseSalary;
            return _salaryBands[_salaryBands.Length - 1].baseSalary;
        }

        // ── Transfer window actions ───────────────────────────────────────────

        /// <summary>
        /// Opens the transfer window (call at end of season).
        /// Expired contracts are cleared; drivers become free agents.
        /// </summary>
        public void OpenTransferWindow()
        {
            State.IsOpen = true;
            // Remove expired contracts
            State.Contracts.RemoveAll(c => c.IsExpired(State.Season));
        }

        /// <summary>
        /// Closes the transfer window (call before new season starts).
        /// Any unsigned drivers get emergency 1-year contracts.
        /// </summary>
        public void CloseTransferWindow()
        {
            State.IsOpen = false;
            SignEmergencyContracts();
        }

        /// <summary>
        /// Advances the season counter by one year.
        /// Call after the season is complete.
        /// </summary>
        public void AdvanceSeason()
        {
            State.Season++;
            OpenTransferWindow();
        }

        // ── Player offer ──────────────────────────────────────────────────────

        /// <summary>
        /// Makes an offer to sign a driver to the player's team.
        /// Returns an error string if the offer is not valid, otherwise null.
        /// The offer is resolved immediately if the driver is a free agent;
        /// contract-buyouts require AI acceptance.
        /// </summary>
        public string MakeOffer(int playerTeamID, int driverID, int salaryMillions, int durationYears)
        {
            if (!State.IsOpen) return "Transfer window is closed.";

            var driver   = GameDataFactory.GetDriver(driverID);
            if (driver == null) return "Driver not found.";

            var mv = GetMarketValue(driverID);
            if (!mv.IsMarketAvailable) return $"Driver is under contract (has veto clause).";

            if (salaryMillions < mv.MinAcceptableMillions)
                return $"Offer too low — driver minimum is €{mv.MinAcceptableMillions}M.";

            // Check team budget
            var teamContracts = State.GetTeamContracts(playerTeamID);
            if (teamContracts.Count >= 2)
                return "Team already has 2 signed drivers.";

            int remaining = State.RemainingBudget(playerTeamID);
            if (salaryMillions > remaining)
                return $"Insufficient budget — €{remaining}M remaining, €{salaryMillions}M requested.";

            // Calculate acceptance probability
            float accept = AcceptanceProbability(driverID, playerTeamID, salaryMillions);
            bool accepted = _rng.NextDouble() < accept;

            var offer = new TransferOffer
            {
                OfferingTeamID         = playerTeamID,
                DriverID               = driverID,
                ProposedSalaryMillions = salaryMillions,
                ProposedDurationYears  = durationYears,
                IsPlayerOffer          = true,
                IsAccepted             = accepted,
                IsDeclined             = !accepted,
                DeclineReason          = accepted ? null : "Driver chose a different offer.",
            };

            State.ClosedOffers.Add(offer);

            if (accepted)
            {
                var contract = new DriverContract
                {
                    DriverID       = driverID,
                    TeamID         = playerTeamID,
                    DurationYears  = durationYears,
                    SalaryMillions = salaryMillions,
                    SeasonStart    = State.Season,
                    SeasonEnd      = State.Season + durationYears,
                    HasVetoClause  = salaryMillions >= 25,
                };
                State.Contracts.Add(contract);

                if (!State.TeamSpent.ContainsKey(playerTeamID)) State.TeamSpent[playerTeamID] = 0;
                State.TeamSpent[playerTeamID] += salaryMillions;
            }

            return null; // No error
        }

        // ── AI transfer simulation ────────────────────────────────────────────

        /// <summary>
        /// Simulates AI team transfer activity for one transfer window.
        /// Call after OpenTransferWindow() and before CloseTransferWindow().
        /// </summary>
        public List<TransferOffer> RunAITransfers()
        {
            var completed = new List<TransferOffer>();

            // Collect free agents
            var freeAgents = GameDataFactory.Drivers
                .Where(d => State.GetDriverContract(d.id) == null)
                .OrderByDescending(d => (d.pace + d.racecraft) / 2)
                .ToList();

            // AI teams with open seats bid on free agents
            foreach (var team in GameDataFactory.Teams.OrderByDescending(t => t.carPerformance))
            {
                var teamContracts = State.GetTeamContracts(team.id);
                int openSeats     = 2 - teamContracts.Count;
                if (openSeats <= 0) continue;

                int budget = State.RemainingBudget(team.id);

                for (int seat = 0; seat < openSeats && freeAgents.Count > 0; seat++)
                {
                    // Pick best affordable candidate
                    var candidate = freeAgents
                        .FirstOrDefault(d => CalculateBaseSalary(d) <= budget);
                    if (candidate == null) break;

                    int salary   = (int)(CalculateBaseSalary(candidate) * (0.90f + (float)_rng.NextDouble() * 0.20f));
                    int duration = _rng.Next(1, 4);

                    float accept = AcceptanceProbability(candidate.id, team.id, salary);
                    bool  ok     = _rng.NextDouble() < accept;

                    var offer = new TransferOffer
                    {
                        OfferingTeamID         = team.id,
                        DriverID               = candidate.id,
                        ProposedSalaryMillions = salary,
                        ProposedDurationYears  = duration,
                        IsPlayerOffer          = false,
                        IsAccepted             = ok,
                        IsDeclined             = !ok,
                    };
                    completed.Add(offer);
                    State.ClosedOffers.Add(offer);

                    if (ok)
                    {
                        var contract = new DriverContract
                        {
                            DriverID       = candidate.id,
                            TeamID         = team.id,
                            DurationYears  = duration,
                            SalaryMillions = salary,
                            SeasonStart    = State.Season,
                            SeasonEnd      = State.Season + duration,
                            HasVetoClause  = salary >= 25,
                        };
                        State.Contracts.Add(contract);
                        if (!State.TeamSpent.ContainsKey(team.id)) State.TeamSpent[team.id] = 0;
                        State.TeamSpent[team.id] += salary;
                        budget -= salary;
                        freeAgents.Remove(candidate);
                    }
                }
            }

            return completed;
        }

        // ── Helpers ───────────────────────────────────────────────────────────

        private float AcceptanceProbability(int driverID, int teamID, int salaryMillions)
        {
            var  driver  = GameDataFactory.GetDriver(driverID);
            var  team    = GameDataFactory.GetTeam(teamID);
            if  (driver == null || team == null) return 0f;

            var  mv      = GetMarketValue(driverID);
            if  (mv == null) return 0f;

            // Base: does the salary meet expectations?
            float salaryRatio = (float)salaryMillions / mv.BaseValueMillions;
            float base_       = Math.Min(salaryRatio - 0.5f, 1.0f); // 0 if half, 1 if full market

            // Team prestige bonus: top teams more attractive
            float prestigeBonus = (team.carPerformance - 76f) / (97f - 76f) * 0.20f;

            // Competition: other interested teams reduce acceptance
            float rivalry = mv.InterestingTeamCount * 0.05f;

            return Math.Max(0f, Math.Min(1f, base_ + prestigeBonus - rivalry));
        }

        private void SignEmergencyContracts()
        {
            foreach (var driver in GameDataFactory.Drivers)
            {
                if (State.GetDriverContract(driver.id) != null) continue;

                // Emergency 1-year deal at 80% market value
                int salary = (int)(CalculateBaseSalary(driver) * 0.80f);
                State.Contracts.Add(new DriverContract
                {
                    DriverID      = driver.id,
                    TeamID        = driver.teamID,  // Return to home team
                    DurationYears = 1,
                    SalaryMillions = salary,
                    SeasonStart   = State.Season,
                    SeasonEnd     = State.Season + 1,
                    HasVetoClause = false,
                });
            }
        }

        // ── Display helpers ───────────────────────────────────────────────────

        /// <summary>Returns all free agents (drivers with no current contract).</summary>
        public List<DriverInfo> GetFreeAgents()
        {
            return GameDataFactory.Drivers
                .Where(d => State.GetDriverContract(d.id) == null)
                .OrderByDescending(d => (d.pace + d.racecraft) / 2)
                .ToList();
        }

        /// <summary>
        /// Returns drivers whose contracts expire at end of the current season.
        /// Useful for planning ahead.
        /// </summary>
        public List<DriverInfo> GetExpiringContracts(int currentSeason)
        {
            var expiringIDs = State.Contracts
                .Where(c => c.SeasonEnd == currentSeason + 1)
                .Select(c => c.DriverID)
                .ToHashSet();

            return GameDataFactory.Drivers
                .Where(d => expiringIDs.Contains(d.id))
                .ToList();
        }
    }
}

File: autosport-manager-2027/src/core/driver_market.py
"""Driver transfer market: contracts, competitor interest, salary budgets."""

from __future__ import annotations
import random
from dataclasses import dataclass, field
from typing import Optional


# ─── Contract ─────────────────────────────────────────────────────────────────

@dataclass
class DriverContract:
    driver_id:       int
    team_id:         int
    duration_years:  int
    salary_m:        int    # Annual salary in millions €
    season_start:    int
    season_end:      int    # Exclusive
    has_veto:        bool = False

    def is_expired(self, season: int) -> bool:
        return season >= self.season_end

    def years_remaining(self, season: int) -> int:
        return max(0, self.season_end - season)

    def describe(self, season: int) -> str:
        rem   = self.years_remaining(season)
        exp   = "Expiring" if rem == 0 else f"{rem}yr left"
        veto  = " [VETO]" if self.has_veto else ""
        return f"{exp}, €{self.salary_m}M/yr{veto}"


# ─── Transfer offer ───────────────────────────────────────────────────────────

@dataclass
class TransferOffer:
    offering_team_id: int
    driver_id:        int
    salary_m:         int
    duration_years:   int
    is_player_offer:  bool = False
    is_accepted:      bool = False
    is_declined:      bool = False
    decline_reason:   str  = ""


# ─── Market value ─────────────────────────────────────────────────────────────

@dataclass
class DriverMarketValue:
    driver_id:        int
    base_salary_m:    int
    min_salary_m:     int     # Will not sign below this
    is_available:     bool    # No current contract, or expiring
    interest_count:   int     # Number of teams pursuing


# ─── Salary calculation ───────────────────────────────────────────────────────

_SALARY_BANDS: list[tuple[int, int]] = [
    (95, 45),  # Verstappen-tier
    (90, 30),  # Hamilton/Leclerc-tier
    (85, 20),  # Race winner
    (80, 14),  # Mid-field regular
    (75, 10),  # Lower mid-field
    (0,   6),  # Rookie/reserve
]


def _base_salary(driver) -> int:
    avg = (driver.pace + driver.racecraft) // 2
    for min_rating, salary in _SALARY_BANDS:
        if avg >= min_rating:
            return salary
    return 6


# ─── Team budget tiers ────────────────────────────────────────────────────────

_TEAM_BUDGETS: dict[str, int] = {
    "top":   90,   # Combined two-driver budget €M
    "mid":   60,
    "lower": 35,
}


def _team_tier(car_performance: int) -> str:
    if car_performance >= 93:
        return "top"
    if car_performance >= 83:
        return "mid"
    return "lower"


# ─── DriverMarket ─────────────────────────────────────────────────────────────

class DriverMarket:
    """
    Manages driver contracts and the seasonal transfer window.

    At season end:
      market.advance_season()      # opens window, clears expired contracts
      market.run_ai_transfers()    # AI teams sign free agents
      market.close_transfer_window()  # emergency contracts for unsigned drivers

    Player action:
      err = market.make_offer(player_team_id, driver_id, salary, years)
      if err: print("Offer rejected:", err)
    """

    def __init__(self, season: int = 2027, drivers: list = None, teams: list = None,
                 rng: random.Random = None):
        self._rng       = rng or random.Random()
        self.season     = season
        self.is_open    = False

        self._drivers   = {d.id: d for d in (drivers or [])}
        self._teams     = {t.id: t for t in (teams or [])}

        self.contracts: list[DriverContract]  = []
        self.closed_offers: list[TransferOffer] = []

        self.team_budgets: dict[int, int] = {}   # max combined salary
        self.team_spent:   dict[int, int] = {}   # currently committed

        if drivers and teams:
            self._init_contracts()
            self._init_budgets()

    # ── Initialisation ────────────────────────────────────────────────────────

    def _init_contracts(self) -> None:
        for driver in self._drivers.values():
            salary   = _base_salary(driver)
            duration = self._rng.randint(1, 3)
            contract = DriverContract(
                driver_id=driver.id,
                team_id=driver.team_id,
                duration_years=duration,
                salary_m=salary,
                season_start=self.season,
                season_end=self.season + duration,
                has_veto=(salary >= 25),
            )
            self.contracts.append(contract)

    def _init_budgets(self) -> None:
        for team in self._teams.values():
            tier = _team_tier(team.car_performance)
            self.team_budgets[team.id] = _TEAM_BUDGETS[tier]
            self.team_spent[team.id]   = sum(
                c.salary_m for c in self.contracts if c.team_id == team.id)

    # ── Lookup helpers ────────────────────────────────────────────────────────

    def get_contract(self, driver_id: int) -> Optional[DriverContract]:
        return next((c for c in self.contracts if c.driver_id == driver_id), None)

    def get_team_contracts(self, team_id: int) -> list[DriverContract]:
        return [c for c in self.contracts if c.team_id == team_id]

    def remaining_budget(self, team_id: int) -> int:
        return self.team_budgets.get(team_id, 20) - self.team_spent.get(team_id, 0)

    def market_value(self, driver_id: int) -> Optional[DriverMarketValue]:
        driver = self._drivers.get(driver_id)
        if driver is None:
            return None
        base    = _base_salary(driver)
        contract = self.get_contract(driver_id)
        free     = contract is None or contract.is_expired(self.season)

        # Count AI teams interested (have budget, have an open seat)
        interest = sum(
            1 for t in self._teams.values()
            if (t.id != getattr(driver, "team_id", -1)
                and self.remaining_budget(t.id) >= int(base * 0.8)
                and len(self.get_team_contracts(t.id)) < 2)
        )

        return DriverMarketValue(
            driver_id=driver_id,
            base_salary_m=base,
            min_salary_m=int(base * 0.75),
            is_available=free,
            interest_count=min(interest, 3),
        )

    # ── Transfer window control ───────────────────────────────────────────────

    def open_transfer_window(self) -> None:
        self.is_open = True
        self.contracts = [c for c in self.contracts if not c.is_expired(self.season)]
        # Recalculate spent after removals
        for team_id in self._teams:
            self.team_spent[team_id] = sum(
                c.salary_m for c in self.contracts if c.team_id == team_id)

    def close_transfer_window(self) -> None:
        self.is_open = False
        self._sign_emergency_contracts()

    def advance_season(self) -> None:
        self.season += 1
        self.open_transfer_window()

    # ── Player offer ──────────────────────────────────────────────────────────

    def make_offer(self, player_team_id: int, driver_id: int,
                   salary_m: int, duration_years: int) -> Optional[str]:
        """
        Makes a player offer to sign a driver.
        Returns None on success, or an error string on failure.
        """
        if not self.is_open:
            return "Transfer window is closed."

        driver = self._drivers.get(driver_id)
        if driver is None:
            return "Driver not found."

        mv = self.market_value(driver_id)
        if not mv or not mv.is_available:
            return "Driver is under contract."

        if salary_m < mv.min_salary_m:
            return f"Offer too low — driver minimum is €{mv.min_salary_m}M."

        if len(self.get_team_contracts(player_team_id)) >= 2:
            return "Team already has two drivers signed."

        if salary_m > self.remaining_budget(player_team_id):
            rem = self.remaining_budget(player_team_id)
            return f"Insufficient budget — €{rem}M remaining."

        # Roll for acceptance
        accept_p = self._acceptance_probability(driver_id, player_team_id, salary_m)
        accepted = self._rng.random() < accept_p

        offer = TransferOffer(
            offering_team_id=player_team_id,
            driver_id=driver_id,
            salary_m=salary_m,
            duration_years=duration_years,
            is_player_offer=True,
            is_accepted=accepted,
            is_declined=not accepted,
            decline_reason="" if accepted else "Driver chose a different offer.",
        )
        self.closed_offers.append(offer)

        if accepted:
            self._sign_contract(driver_id, player_team_id, salary_m, duration_years)

        return None  # success

    # ── AI transfers ──────────────────────────────────────────────────────────

    def run_ai_transfers(self) -> list[TransferOffer]:
        """Simulates AI team transfer activity. Returns completed offer list."""
        completed: list[TransferOffer] = []

        # Free agents, ordered by quality
        free_agents = [
            d for d in self._drivers.values()
            if self.get_contract(d.id) is None
        ]
        free_agents.sort(key=lambda d: -(d.pace + d.racecraft))

        # Teams with open seats, ordered by car performance (best team first)
        teams_with_seats = [
            t for t in sorted(self._teams.values(), key=lambda t: -t.car_performance)
            if len(self.get_team_contracts(t.id)) < 2
        ]

        for team in teams_with_seats:
            open_seats = 2 - len(self.get_team_contracts(team.id))
            budget     = self.remaining_budget(team.id)

            for _ in range(open_seats):
                if not free_agents:
                    break
                candidate = next(
                    (d for d in free_agents if _base_salary(d) <= budget), None)
                if candidate is None:
                    break

                salary   = int(_base_salary(candidate) * (0.90 + self._rng.random() * 0.20))
                duration = self._rng.randint(1, 3)

                accept_p = self._acceptance_probability(candidate.id, team.id, salary)
                accepted = self._rng.random() < accept_p

                offer = TransferOffer(
                    offering_team_id=team.id,
                    driver_id=candidate.id,
                    salary_m=salary,
                    duration_years=duration,
                    is_accepted=accepted,
                    is_declined=not accepted,
                )
                completed.append(offer)
                self.closed_offers.append(offer)

                if accepted:
                    self._sign_contract(candidate.id, team.id, salary, duration)
                    budget -= salary
                    free_agents.remove(candidate)

        return completed

    # ── Display helpers ───────────────────────────────────────────────────────

    def free_agents(self) -> list:
        return sorted(
            [d for d in self._drivers.values() if self.get_contract(d.id) is None],
            key=lambda d: -(d.pace + d.racecraft)
        )

    def expiring_contracts(self) -> list:
        """Returns drivers whose contracts expire at end of this season."""
        expiring_ids = {c.driver_id for c in self.contracts
                       if c.season_end == self.season + 1}
        return [d for d in self._drivers.values() if d.id in expiring_ids]

    # ── Internal helpers ──────────────────────────────────────────────────────

    def _acceptance_probability(self, driver_id: int, team_id: int, salary_m: int) -> float:
        driver = self._drivers.get(driver_id)
        team   = self._teams.get(team_id)
        if driver is None or team is None:
            return 0.0
        mv = self.market_value(driver_id)
        if mv is None:
            return 0.0

        salary_ratio    = salary_m / mv.base_salary_m if mv.base_salary_m else 1.0
        base            = min(salary_ratio - 0.5, 1.0)  # 0 if half salary, 1 if full
        prestige_bonus  = (team.car_performance - 76) / (97 - 76) * 0.20
        rivalry_penalty = mv.interest_count * 0.05
        return max(0.0, min(1.0, base + prestige_bonus - rivalry_penalty))

    def _sign_contract(self, driver_id: int, team_id: int,
                       salary_m: int, duration_years: int) -> None:
        contract = DriverContract(
            driver_id=driver_id,
            team_id=team_id,
            duration_years=duration_years,
            salary_m=salary_m,
            season_start=self.season,
            season_end=self.season + duration_years,
            has_veto=(salary_m >= 25),
        )
        self.contracts.append(contract)
        self.team_spent[team_id] = self.team_spent.get(team_id, 0) + salary_m

    def _sign_emergency_contracts(self) -> None:
        for driver in self._drivers.values():
            if self.get_contract(driver.id) is not None:
                continue
            salary = int(_base_salary(driver) * 0.80)
            self._sign_contract(driver.id, driver.team_id, salary, 1)
